extmod/network_lwip.c

```diff
@@ -90,6 +99,279 @@ mp_obj_t mod_network_nic_ifconfig(struct netif *netif, size_t n_args, const mp_o
+                struct dhcp6 *dhcp = netif_dhcp6_data(netif);
+                return mp_obj_new_bool(dhcp != NULL && dhcp->state != DHCP6_STATE_OFF);
+            }
+            #endif
+            #if LWIP_IPV6_AUTOCONFIG
+            case MP_QSTR_autoconf6: {
+                return netif->ip6_autoconfig_enabled ?  mp_const_true : mp_const_false;
+            }
+            #endif
+            case MP_QSTR_addr4: {
+                mp_obj_t tuple[2] = {
+                    netutils_format_ipv4_addr((uint8_t *)&netif->ip_addr, NETUTILS_BIG),
+                    netutils_format_ipv4_addr((uint8_t *)&netif->netmask, NETUTILS_BIG),
+                };
+                return mp_obj_new_tuple(2, tuple);
+            }
+            #if LWIP_IPV6
+            case MP_QSTR_addr6: {
+                mp_obj_t addrs[LWIP_IPV6_NUM_ADDRESSES];
+                size_t n_addrs = 0;
+                for (int i = 0; i < LWIP_IPV6_NUM_ADDRESSES; i++) {
+                    if (ip6_addr_isvalid(netif_ip6_addr_state(netif, i))) {
+                        char addr_str[IPADDR_STRLEN_MAX];
+                        ipaddr_ntoa_r(netif_ip_addr6(netif, i), addr_str, sizeof(addr_str));
+                        mp_obj_t tuple[4] = {
+                            mp_obj_new_str(addr_str, strlen(addr_str)),
+                            MP_OBJ_NEW_SMALL_INT(netif_ip6_addr_state(netif, i)),
+                            MP_OBJ_NEW_SMALL_INT(netif_ip6_addr_pref_life(netif, i)), // preferred
+                            MP_OBJ_NEW_SMALL_INT(netif_ip6_addr_valid_life(netif, i))
+                        };
+                        addrs[n_addrs++] = mp_obj_new_tuple(4, tuple);
+                    }
+                }
+                return mp_obj_new_list(n_addrs, addrs);
+            }
+            #endif
+            case MP_QSTR_gw4: {
+                return netutils_format_ipv4_addr((uint8_t *)&netif->gw, NETUTILS_BIG);
+            }
+            default: {
+                mp_raise_ValueError(MP_ERROR_TEXT("unexpected key"));
+                break;
+            }
+        }
+        return mp_const_none;
+    } else {
+        // Set config value(s)
+        if (n_args != 0) {
+            mp_raise_TypeError(MP_ERROR_TEXT("can't specify pos and kw args"));
+        }
+
+        for (size_t i = 0; i < kwargs->alloc; ++i) {
+            if (MP_MAP_SLOT_IS_FILLED(kwargs, i)) {
+                mp_map_elem_t *e = &kwargs->table[i];
+                switch (mp_obj_str_get_qstr(e->key)) {
+                    case MP_QSTR_dhcp4: {
+                        if (mp_obj_is_true(e->value)) {
+                            if (dhcp_supplied_address(netif)) {
+                                dhcp_renew(netif);
+                            } else {
+                                dhcp_release_and_stop(netif);
+                                dhcp_start(netif);
+                            }
+                            uint32_t start = mp_hal_ticks_ms();
```

Comment:
From talking to @dpgeorge and @projectgus about this:

The only other thing in network/nic config that blocks right now is explicitly doing `nic.ifconfig("dhcp")`. Note that the initial dhcp fetch doesn't block. Seeing as we're deprecating `ifconfig`, I think we should just declare that the `network` module is entirely non-blocking.

This is exactly your proposed "Option 1". However, we need to be able from Python to poll for an autoconf/dhcp address, i.e. an equivalent of `dhcp_supplied_address` or `!netif_ip6_addr_isstatic`. I think that means we need to add another field to the return tuple of `addr4` and `addr6` ? Or other suggestions welcome... `nic.ipconfig("has_autoconf6")` / `nic.ipconfig("has_dhcp4")` ? Not sure...

In the future we should consider adding callbacks (or a `irq` method, similar to the machine APIs) -- this is a long-requested feature.